sql/src/main/java/org/apache/druid/sql/calcite/rel/DruidQuery.java

```diff
@@ -1442,12 +1445,30 @@ private WindowOperatorQuery toWindowQuery()
       return null;
     }
 
+    // all virtual cols are needed - these columns are only referenced from the aggregates
+    VirtualColumns virtualColumns = virtualColumnRegistry.build(Collections.emptySet());
+    final List<OperatorFactory> operators;
+
+    if (!virtualColumns.isEmpty()) {
```

Comment:
This is a peeve of mine, but you are doing a `!` here in an `if/else`.  You can make the logic positive and just move the if stuff to the else stuff.  I find that every negation in a boolean operation halves the readability of the code (you have to take a moment to try to figure out what it means), so try to keep all boolean expressions as "positive" as possible.

Reply:
I'm not sure but that `!` may even worth an extra point of complexity :D
I've negated the if :D